### PR TITLE
[PyTorch] Allow F16 GDN inside FP8 autocast

### PR DESCRIPTION
**Pull request by layalir**
# Description

GatedDeltaNetAttention currently rejects any enclosing FP8 autocast or calibration context, preventing FP16/BF16 GDN from being composed with MXFP8 linear layers.

Add `NVTE_LA_FP8_RECIPE=F16` to run GDN in its input FP16/BF16 dtype while retaining the enclosing linear-layer recipe. Set the variable before importing Transformer Engine. Linear attention has its own override, independent of `NVTE_DPA_FP8_RECIPE`.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Changes

- Disable quantization and calibration around the complete GDN forward lifecycle, including metadata preparation and checkpoint dispatch; restore the enclosing context on normal and exceptional exits.
- Preserve FP32 gate/state requirements and the existing rejection when the override is absent or unsupported.
- Document the variable and add regression coverage for calibration, context restoration, and FP16/BF16 GDN between MXFP8 linear layers with eager and checkpointed backward.

## Validation

On GB300, the following selection passed **16 tests with zero failures or skips**, including forward comparison with the reference recurrence and finite gradients:

```bash
NVTE_GDN_TEST_REQUIRED=1 CUDA_DEVICE_MAX_CONNECTIONS=1 \
python -m pytest -q tests/pytorch/attention/test_gdn_attention.py \
  -k 'f16_override or rejects_fp8_autocast'
```

Also completed ten synthetic policy-training steps on eight GB300 GPUs. Runtime observation confirmed BF16 GDN/GQA and MXFP8 expert grouped-GEMM forward, input-gradient, and weight-gradient operations on every rank. This verifies mixed-precision integration, not convergence or performance.

# Checklist

- [x] Read the contributing guidelines; both commits are signed off.
- [x] Functionality, explanatory comments, documentation, and regression tests are included.
- [x] Targeted GPU regression tests pass.
- [ ] Full repository L0 test and lint suites (not run; targeted validation is described above).

### CODE REVIEW

**Review by ksivaman — CHANGES_REQUESTED** [review]
This can be closed in favor of #3521 which removes this restriction and achieves the same outcome. I would avoid introducing environment variables here, especially as multiple recipes are not supported for LA as of yet.

**Comment by layalir** [comment]
Makes sense. Closing.
